README
==============================================
    Speech recognition script for Asterisk
==============================================

This script makes use of Google's Cloud Speech API in order to render speech
to text and return it back to the dialplan as an asterisk channel variable.

------------
Requirements
------------
Perl          The Perl Programming Language
perl-libwww   The World-Wide Web library for Perl
perl-libjson  Module for manipulating JSON-formatted data
IO-Socket-SSL Perl module that implements an interface to SSL sockets.
flac          Free Lossless Audio Codec

Cloud Speech API key from Google (https://cloud.google.com/speech).
Internet access in order to contact Google and get the speech data.

------------
Installation
------------
To install copy speech-recog.agi to your agi-bin directory.
Usually this is /var/lib/asterisk/agi-bin/
To make sure check your /etc/asterisk/asterisk.conf file

-----
Usage
-----
agi(speech-recog.agi,[lang],[timeout],[intkey],[NOBEEP])
Records from the current channel until 2 seconds of silence are detected
(this can be set by the user by the 'timeout' argument, -1 for no timeout) or the
interrupt key (# by default) is pressed. If NOBEEP is set, no beep sound is played
back to the user to indicate the start of the recording.
The recorded sound is send over to Google speech recognition service and the
returned text string is assigned as the value of the channel variable 'utterance'.
The scripts sets the following channel variables:

utterance  : The generated text string.
confidence : A value between 0 and 1 indicating the probability of a correct recognition.
             Values bigger than 0.95 usually mean that the resulted text is correct.

In case of an unexpected error both these variables are set to '-1'.

--------
Examples
--------
sample dialplan code for your extensions.conf

;Simple speech recognition
exten => 1234,1,Answer()
exten => 1234,n,agi(speech-recog.agi,en-US)
exten => 1234,n,Verbose(1,The text you just said is: ${utterance})
exten => 1234,n,Verbose(1,The probability to be right is: ${confidence})
exten => 1234,n,Hangup()

;Speech recognition demo also using googletts.agi for text to speech synthesis:
exten => 1235,1,Answer()
exten => 1235,n,agi(googletts.agi,"Say something in English, when done press the pound key.",en)
exten => 1235,n(record),agi(speech-recog.agi,en-US)
exten => 1235,n,Verbose(1,Script returned: ${confidence} , ${utterance})

;Check the probability of a successful recognition:
exten => 1235,n,GotoIf($["${confidence}" > "0.8"]?playback:retry)

;Playback the text
exten => 1235,n(playback),agi(googletts.agi,"The text you just said was...",en)
exten => 1235,n,agi(googletts.agi,"${utterance}",en)
exten => 1235,n,goto(end)

;Retry in case speech recognition wasn't successful:
exten => 1235,n(retry),agi(googletts.agi,"Can you please repeat more clearly?",en)
exten => 1235,n,goto(record)

exten => 1235,n(fail),agi(googletts.agi,"Failed to get speech data.",en)
exten => 1235,n(end),Hangup()

;Voice dialing example
exten => 1236,1,Answer()
exten => 1236,n,agi(googletts.agi,"PLease say the number you want to dial.",en)
exten => 1236,n(record),agi(speech-recog.agi,en-US)
exten => 1236,n,GotoIf($["${confidence}" > "0.8"]?success:retry)

exten => 1236,n(success),goto(${utterance},1)

exten => 1236,n(retry),agi(googletts.agi,"Can you please repeat?",en)
exten => 1236,n,goto(record)

-------------------
Supported Languages
-------------------
"af-ZA" Afrikaans (Suid-Afrika)
"id-ID" Bahasa Indonesia (Indonesia)
"ms-MY" Bahasa Melayu (Malaysia)
"ca-ES" Català (Espanya)
"cs-CZ" Čeština (Česká republika)
"da-DK" Dansk (Danmark)
"de-DE" Deutsch (Deutschland)
"en-AU" English (Australia)
"en-CA" English (Canada)
"en-GB" English (Great Britain)
"en-IN" English (India)
"en-IE" English (Ireland)
"en-NZ" English (New Zealand)
"en-PH" English (Philippines)
"en-ZA" English (South Africa)
"en-US" English (United States)
"es-AR" Español (Argentina)
"es-BO" Español (Bolivia)
"es-CL" Español (Chile)
"es-CO" Español (Colombia)
"es-CR" Español (Costa Rica)
"es-EC" Español (Ecuador)
"es-SV" Español (El Salvador)
"es-ES" Español (España)
"es-US" Español (Estados Unidos)
"es-GT" Español (Guatemala)
"es-HN" Español (Honduras)
"es-MX" Español (México)
"es-NI" Español (Nicaragua)
"es-PA" Español (Panamá)
"es-PY" Español (Paraguay)
"es-PE" Español (Perú)
"es-PR" Español (Puerto Rico)
"es-DO" Español (República Dominicana)
"es-UY" Español (Uruguay)
"es-VE" Español (Venezuela)
"eu-ES" Euskara (Espainia)
"fil-PH" Filipino (Pilipinas)
"fr-FR" Français (France)
"gl-ES" Galego (España)
"hr-HR" Hrvatski (Hrvatska)
"zu-ZA" IsiZulu (Ningizimu Afrika)
"is-IS" Íslenska (Ísland)
"it-IT" Italiano (Italia)
"lt-LT" Lietuvių (Lietuva)
"hu-HU" Magyar (Magyarország)
"nl-NL" Nederlands (Nederland)
"nb-NO" Norsk bokmål (Norge)
"pl-PL" Polski (Polska)
"pt-BR" Português (Brasil)
"pt-PT" Português (Portugal)
"ro-RO" Română (România)
"sk-SK" Slovenčina (Slovensko)
"sl-SI" Slovenščina (Slovenija)
"fi-FI" Suomi (Suomi)
"sv-SE" Svenska (Sverige)
"vi-VN" Tiếng Việt (Việt Nam)
"tr-TR" Türkçe (Türkiye)
"el-GR" Ελληνικά (Ελλάδα)
"bg-BG" Български (България)
"ru-RU" Русский (Россия)
"sr-RS" Српски (Србија)
"uk-UA" Українська (Україна)
"he-IL" עברית (ישראל)
"ar-IL" العربية (إسرائيل)
"ar-JO" العربية (الأردن)
"ar-AE" العربية (الإمارات)
"ar-BH" العربية (البحرين)
"ar-DZ" العربية (الجزائر)
"ar-SA" العربية (السعودية)
"ar-IQ" العربية (العراق)
"ar-KW" العربية (الكويت)
"ar-MA" العربية (المغرب)
"ar-TN" العربية (تونس)
"ar-OM" العربية (عُمان)
"ar-PS" العربية (فلسطين)
"ar-QA" العربية (قطر)
"ar-LB" العربية (لبنان)
"ar-EG" العربية (مصر)
"fa-IR" فارسی (ایران)
"hi-IN" हिन्दी (भारत)
"th-TH" ไทย (ประเทศไทย)
"ko-KR" 한국어 (대한민국)
"cmn-Hant-TW" 國語 (台灣)
"yue-Hant-HK" 廣東話 (香港)
"ja-JP" 日本語（日本）
"cmn-Hans-HK" 普通話 (香港)
"cmn-Hans-CN" 普通话 (中国大陆)

-----------------------
Security Considerations
-----------------------
This script contacts Google servers in order send the recorded voice data and get back
the resulted text. The script uses TLS by default to encrypt all the traffic between
your PBX and Google servers so no 3rd party can eavesdrop your communication, but your
voice data will be available to Google under a not yet defined policy.

------------
Tiny version
------------
The '-tiny' suffixed scripts use the HTTP::Tiny perl module instead of LWP::UserAgent and
JSON::Tiny instead of JSON. This makes them a lot faster when run in small embedded systems
or boards like the Raspberry pi. They can be used as an in-place replacement of the normal
scripts and expose the same interface/cli args. To use them just make sure
you have HTTP::Tiny and JSON::Tiny installed.

-------
License
-------
The speech-recog script for asterisk is distributed under the GNU General Public
License v2. See COPYING for details.

--------
Homepage
--------
http://zaf.github.com/asterisk-speech-recog/


ChangeLog
2017-10-20 Lefteris Zafiris <zaf@fastmail.com> - 0.8
	Added support for the Cloud SPeech API. Support for
	the old Speech API removed.

2014-05-28 Lefteris Zafiris <zaf@fastmail.com> - 0.7
	Added support for version 2 of Speech API.

2012-09-30 Lefteris Zafiris <zaf@fastmail.com> - 0.6
	Added automatic audio format detection. Enabled SSL by default.
	Proxy support. Added profanities filtering option.
	Added 'maxresults' and 'grammar' options.
	CLI app can now accept flac, speex or wav files as input.

2012-02-28 Lefteris Zafiris <zaf@fastmail.com> - 0.6-beta1
	Added support for speex encoded voice data.
	Works only with patched speex encoder that
	supports MIME "x-speex-with-header-byte":
	https://github.com/zaf/Speex-with-header-bytes
	Speex support is experimental and detection rates are
	lower compared to flac encoded data.

2012-02-01 Lefteris Zafiris <zaf@fastmail.com> - 0.5
	Added SSL support. Interrupt keys can now be defined
	by the user. Added (optional) Beep indication before recording.
	Recording time is now conrtolled by silence detection. There is no
	absolute timeout, the recording will terminate after a
	user defined (3 sec by default) period of silence, or the pressing
	of the interrupt key(s).
	Added a sample AGI script that submits queries to WolframAlpha and
	sets the answer as a dialplan variable.

2012-01-10 Lefteris Zafiris <zaf@fastmail.com> - 0.4
	Added option for user defined timeout.
	Remove spaces between digits in returning text string.
	Documentation updates.

2012-01-05 Lefteris Zafiris <zaf@fastmail.com> - 0.3
	Replaced sox with flac encoder for compatibility
	with older distros (RHEL 5). Sox can be enabled
	by the user.
	When sox is in use it normalizes the audio, removes
	low frequency noise and possible DC offset.

2012-01-04 Lefteris Zafiris <zaf@fastmail.com> - 0.2
	Added option for setting the speech language.
	More debugging messages added.
	Recording timeout set to 15 seconds.
	Documentation updates and diaplan examples.

2011-12-29 Lefteris Zafiris <zaf@fastmail.com> - 0.1
	Initial release 0.1


wolfram/wolfram.pl
#!/usr/bin/env perl

#
# Submit question to WolframAlpha and print the answer.
#
# Copyright (C) 2012, Lefteris Zafiris <zaf.000@gmail.com>
#
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2. See the COPYING file
# at the top of the source tree.
#
# To use this script you must get an app ID from http://products.wolframalpha.com/api
#
# **This is still experimental code and the output is far from perfect.**
# If you failed to get any results set $debug = 1 this will display a raw output of the
# data we got from wolfram and might help to find the place where the answer is hiding.
#

use strict;
#use warnings;
use LWP::UserAgent;
use CGI::Util qw(escape);
use XML::Simple;
use YAML;

# Here you can assign your App ID from wolfram #
my $app_id   = "";
my $debug    = 0;
my $url      = "http://api.wolframalpha.com/v2/query";
my $question = escape($ARGV[0]);
my $results  = 0;

if (!@ARGV || $ARGV[0] eq '-h' || $ARGV[0] eq '--help') {
	print "Ask WolframAlpha.\n\n";
	print "Usage:   $0 [question]\n\n";
	print "Example: $0 \"What time is it?\"\n";
	exit;
}
die "You must have an App ID from WolframAlpha to use this script.\n" if (!$app_id);

my $ua = LWP::UserAgent->new;
$ua->agent("Mozilla/5.0 (X11; Linux) AppleWebKit/535.2 (KHTML, like Gecko)");
$ua->timeout(15);
my $ua_request = HTTP::Request->new(
	'GET' => "$url?input=$question&appid=$app_id".
		"&format=plaintext&scantimeout=8&excludepodid=Input&excludepodid=Interpretation"
);
my $ua_response = $ua->request($ua_request);
if (!$ua_response->is_success) {
	print "Failed to contact server.\n";
	exit;
}

my $answer = XMLin($ua_response->content);
if ($answer->{success} eq 'false') {
	print "I don't know how to answer that.\n";
	exit;
}

print "====Raw output:====\n", Dump($answer), "\n===End of raw output.===\n" if ($debug);

foreach (keys %{$answer->{pod}}) {
	if (/subpod/) {
		print "$answer->{pod}{$_}{plaintext}\n";
		$results++;
		last;
	} elsif (/Result|Value/) {
		eval{ print "$answer->{pod}{$_}{subpod}{plaintext}\n"; };
		eval{ print "$answer->{pod}{$_}{subpod}[0]{plaintext}\n"; };
		eval{ print "$answer->{pod}{$_}{subpod}[1]{plaintext}\n"; };
		$results++;
		last;
	} elsif (/Definition:WordData|Basic:ChemicalData|ComparisonAsLength|Comparison|BasicInformation|Basic|Properties/) {
		print "$answer->{pod}{$_}{subpod}{plaintext}\n";
		$results++;
		last;
	} elsif (/NotableFacts:PeopleData|BasicInformation:PeopleData/) {
		print "$answer->{pod}{$_}{subpod}{plaintext}\n";
		$results++;
	} elsif (/WeatherForecast:WeatherData/) {
		print "$answer->{pod}{$_}{subpod}[0]{title} $answer->{pod}{$_}{subpod}[0]{plaintext}\n";
		print "$answer->{pod}{$_}{subpod}[1]{title} $answer->{pod}{$_}{subpod}[1]{plaintext}\n";
		eval{ print "$answer->{pod}{$_}{subpod}[2]{title} $answer->{pod}{$_}{subpod}[2]{plaintext}\n"; };
		$results++;
		last;
	}
}

if (!$results) {
	foreach (keys %{$answer->{pod}}) {
		eval{ print "$_: $answer->{pod}{$_}{subpod}{plaintext}\n"; };
		eval{ print "$_: $answer->{pod}{$_}{subpod}[0]{plaintext}\n"; };
		eval{ print "$_: $answer->{pod}{$_}{subpod}[1]{plaintext}\n"; };
		$results++;
	}
}
print "Failed to get any resutl, possible script bug.\n" if (!$results);


cli/speech-recog-cli.pl
#!/usr/bin/env perl

#
# Render speech to text using Google's Cloud Speech API.
#
# Copyright (C) 2011 - 2016, Lefteris Zafiris <zaf@fastmail.com>
#
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2. See the COPYING file
# at the top of the source tree.
#
# The script takes as input flac or wav files and returns the following values:
# transcript : The recognized text string.
# confidence : A value between 0 and 1 indicating how "confident" the recognition engine
#  feels about the result. Values bigger than 0.95 usually mean that the
#  resulted text is correct.
#

use strict;
use warnings;
use File::Temp qw(tempfile);
use Getopt::Std;
use File::Basename;
use LWP::UserAgent;
use LWP::ConnCache;
use JSON;
use MIME::Base64;

my %options;
my $flac;
my $key;
my $url        = "https://speech.googleapis.com/v1/speech";
my $samplerate = 16000;
my $language   = "en-US";
my $output     = "detailed";
my $results    = 1;
my $pro_filter = "false";
my $error      = 0;

getopts('k:l:o:r:n:fhq', \%options);

VERSION_MESSAGE() if (defined $options{h} || !@ARGV);

parse_options();

my %config = (
	"encoding"         => "FLAC",
	"sampleRateHertz"  => $samplerate,
	"languageCode"     => $language,
	"profanityFilter"  => $pro_filter,
	"maxAlternatives"  => $results,
);

my $ua = LWP::UserAgent->new(ssl_opts => {verify_hostname => 1});
$ua->agent("CLI speech recognition script");
$ua->env_proxy;
$ua->conn_cache(LWP::ConnCache->new());
$ua->timeout(60);

# send each sound file to Google and get the recognition results #
foreach my $file (@ARGV) {
	my ($filename, $dir, $ext) = fileparse($file, qr/\.[^.]*/);
	if ($ext ne ".flac" && $ext ne ".wav") {
		say_msg("Unsupported file-type: $ext");
		++$error;
		next;
	}
	if ($ext eq ".wav") {
		if (($file = encode_flac($file)) eq '-1') {
			++$error;
			next;
		}
	}
	print("Opening $filename\n") if (!defined $options{q});
	my $audio;
	if (open(my $fh, "<", "$file")) {
		$audio = do { local $/; <$fh> };
		close($fh);
	} else {
		say_msg("Cant read file $file");
		++$error;
		next;
	}
	my %audio = ( "content" => encode_base64($audio, "") );
	my %json = (
		"config" => \%config,
		"audio"  => \%audio,
	);
	my $response = $ua->post(
		"$url:recognize?key=$key",
		Content_Type => "application/json",
		Content      => encode_json(\%json),
	);
	if (!$response->is_success) {
		say_msg("Failed to get data for file: $file");
		++$error;
		next;
	}
	if ($output eq "raw") {
		print $response->content;
		next;
	}
	my $jdata = decode_json($response->content);
	if ($output eq "detailed") {
		foreach (@{$jdata->{"results"}[0]->{"alternatives"}}) {
			printf "%-10s : %s\n", "transcript", $_->{"transcript"};
			printf "%-10s : %s\n", "confidence", $_->{"confidence"} if $_->{"confidence"};
		}
	} elsif ($output eq "compact") {
		print $_->{"transcript"}."\n" foreach (@{$jdata->{"results"}[0]->{"alternatives"}});
	}
}

exit(($error) ? 1 : 0);

sub parse_options {
# Command line options parsing #
	if (defined $options{k}) {
	# check API key #
		$key = $options{k};
	} else {
		say_msg("Invalid or missing API key.\n");
		exit 1;
	}
	if (defined $options{l}) {
	# check if language setting is valid #
		if ($options{l} =~ /^[a-z]{2}(-[a-zA-Z]{2,6})?$/) {
			$language = $options{l};
		} else {
			say_msg("Invalid language setting. Using default.\n");
		}
	}
	if (defined $options{o}) {
	# check if output setting is valid #
		if ($options{o} =~ /^(detailed|compact|raw)$/) {
			$output = $options{o};
		} else {
			say_msg("Invalid output formatting setting. Using default.\n");
		}
	}
	if (defined $options{n}) {
	# set number or results #
		$results = $options{n} if ($options{n} =~ /\d+/);
	}
	if (defined $options{r}) {
	# set audio sampling rate #
		$samplerate = $options{r} if ($options{r} =~ /\d+/);
	}
	# set profanity filter #
	$pro_filter = "true" if (defined $options{f});

	return;
}

sub encode_flac {
# Encode file to flac and return the filename #
	my $file   = shift;
	my $tmpdir = "/tmp";
	if (!$flac) {
		$flac   = `/usr/bin/which flac`;
		if (!$flac) {
			say_msg("flac encoder is missing. Aborting.");
			return -1;
		}
		chomp($flac);
	}
	my ($fh, $tmpname) = tempfile(
		"recg_XXXXXX",
		DIR    => $tmpdir,
		SUFFIX => '.flac',
		UNLINK => 1,
	);
	if (system($flac, "-8", "-f", "--totally-silent", "-o", "$tmpname", "$file")) {
		say_msg("$flac failed to encode file");
		return -1;
	}
	return $tmpname;
}

sub say_msg {
# Print messages to user if 'quiet' flag is not set #
	my @message = @_;
	warn @message if (!defined $options{q});
	return;
}

sub VERSION_MESSAGE {
# Help message #
	print "Speech recognition using Google Cloud Speech API.\n\n",
		"Usage: $0 [options] [file(s)]\n\n",
		"Supported options:\n",
		" -k <key>       specify the Speech API key\n",
		" -l <lang>      specify the language to use (default 'en-US')\n",
		" -o <type>      specify the type of output formatting\n",
		"    detailed    print detailed output with info like confidence (default)\n",
		"    compact     print only the transcripted string\n",
		"    raw         raw JSON output\n",
		" -r <rate>      specify the audio sample rate in Hz (default 16000)\n",
		" -n <number>    specify the maximum number of results (default 1)\n",
		" -f             filter out profanities\n",
		" -q             don't print any error messages or warnings\n",
		" -h             this help message\n\n";
	exit(1);
}


samples/wolfram/wolfram-cli.pl
#!/usr/bin/env perl

#
# Submit queries to WolframAlpha and print the answers.
#
# Copyright (C) 2012 - 2014, Lefteris Zafiris <zaf@fastmail.com>
#
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2. See the COPYING file
# at the top of the source tree.
#
# To use this script you must get an app ID from http://products.wolframalpha.com/api
#
# **This is still experimental code and the output is far from perfect.**
# If you failed to get any results set $debug = 1 this will display a raw output of the
# data we got from wolfram and might help to find the place where the answer is hiding.
#

use strict;
#use warnings;
use LWP::UserAgent;
use CGI::Util qw(escape);
use XML::Simple;
use YAML;

# Here you can assign your App ID from wolfram #
my $app_id   = "";
my $debug    = 0;
my $url      = "http://api.wolframalpha.com/v2/query";
my $question = escape($ARGV[0]);
my $results  = 0;

if (!@ARGV || $ARGV[0] eq '-h' || $ARGV[0] eq '--help') {
	print "Ask WolframAlpha.\n\n";
	print "Usage:   $0 [question]\n\n";
	print "Example: $0 \"What time is it?\"\n";
	exit;
}
die "You must have an App ID from WolframAlpha to use this script.\n" if (!$app_id);

my $ua = LWP::UserAgent->new;
$ua->agent("Mozilla/5.0 (X11; Linux) AppleWebKit/535.2 (KHTML, like Gecko)");
$ua->env_proxy;
$ua->timeout(15);
my $ua_request = HTTP::Request->new(
	'GET' => "$url?input=$question&appid=$app_id".
		"&format=plaintext&scantimeout=8&excludepodid=Input&excludepodid=Interpretation"
);
my $ua_response = $ua->request($ua_request);
if (!$ua_response->is_success) {
	print "Failed to contact server.\n";
	exit;
}

my $answer = XMLin($ua_response->content);
if ($answer->{success} eq 'false') {
	print "I don't know how to answer that.\n";
	exit;
}

print "====Raw output:====\n", Dump($answer), "\n===End of raw output.===\n" if ($debug);

foreach (keys %{$answer->{pod}}) {
	if (/subpod/) {
		print "$answer->{pod}{$_}{plaintext}\n";
		$results++;
		last;
	} elsif (/Result|Value/) {
		eval{ print "$answer->{pod}{$_}{subpod}{plaintext}\n"; };
		eval{ print "$answer->{pod}{$_}{subpod}[0]{plaintext}\n"; };
		eval{ print "$answer->{pod}{$_}{subpod}[1]{plaintext}\n"; };
		$results++;
		last;
	} elsif (/Definition:WordData|Basic:ChemicalData|ComparisonAsLength|Comparison|
				BasicInformation|NotableFacts|Basic|Properties/x) {
		print "$answer->{pod}{$_}{subpod}{plaintext}\n";
		$results++;
	} elsif (/WeatherForecast:WeatherData/) {
		print "$answer->{pod}{$_}{subpod}[0]{title} $answer->{pod}{$_}{subpod}[0]{plaintext}\n";
		print "$answer->{pod}{$_}{subpod}[1]{title} $answer->{pod}{$_}{subpod}[1]{plaintext}\n";
		eval{ print "$answer->{pod}{$_}{subpod}[2]{title} $answer->{pod}{$_}{subpod}[2]{plaintext}\n"; };
		$results++;
		last;
	}
}

if (!$results) {
	foreach (keys %{$answer->{pod}}) {
		eval{ print "$_: $answer->{pod}{$_}{subpod}{plaintext}\n"; };
		eval{ print "$_: $answer->{pod}{$_}{subpod}[0]{plaintext}\n"; };
		eval{ print "$_: $answer->{pod}{$_}{subpod}[1]{plaintext}\n"; };
		$results++;
	}
}
print "Failed to get any resutl, possible script bug.\n" if (!$results);


cli/speech-recog-cli-tiny.pl
#!/usr/bin/env perl

#
# Render speech to text using Google's Cloud Speech API.
#
# Copyright (C) 2011 - 2016, Lefteris Zafiris <zaf@fastmail.com>
#
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2. See the COPYING file
# at the top of the source tree.
#
# The script takes as input flac or wav files and returns the following values:
# transcript : The recognized text string.
# confidence : A value between 0 and 1 indicating how "confident" the recognition engine
#  feels about the result. Values bigger than 0.95 usually mean that the
#  resulted text is correct.
#

use strict;
use warnings;
use File::Temp qw(tempfile);
use Getopt::Std;
use File::Basename;
use HTTP::Tiny;
use JSON::Tiny qw(decode_json encode_json);
use MIME::Base64;

my %options;
my $flac;
my $key;
my $url        = "https://speech.googleapis.com/v1/speech";
my $samplerate = 16000;
my $language   = "en-US";
my $output     = "detailed";
my $results    = 1;
my $pro_filter = "false";
my $error      = 0;

getopts('k:l:o:r:n:fhq', \%options);

VERSION_MESSAGE() if (defined $options{h} || !@ARGV);

parse_options();

my %config = (
	"encoding"         => "FLAC",
	"sampleRateHertz"  => $samplerate,
	"languageCode"     => $language,
	"profanityFilter"  => $pro_filter,
	"maxAlternatives"  => $results,
);

my $http = HTTP::Tiny->new(
	agent      => 'CLI speech recognition script',
	timeout    => 60,
	verify_SSL => 1,
);

# send each sound file to google and get the recognition results #
foreach my $file (@ARGV) {
	my ($filename, $dir, $ext) = fileparse($file, qr/\.[^.]*/);
	if ($ext ne ".flac" && $ext ne ".wav") {
		say_msg("Unsupported file type: $ext");
		++$error;
		next;
	}
	if ($ext eq ".wav") {
		if (($file = encode_flac($file)) eq '-1') {
			++$error;
			next;
		}
	}
	print("Opening $filename\n") if (!defined $options{q});
	my $audio;
	if (open(my $fh, "<", "$file")) {
		$audio = do { local $/; <$fh> };
		close($fh);
	} else {
		say_msg("Cant read file $file");
		++$error;
		next;
	}
	my %audio = ( "content" => encode_base64($audio, "") );
	my %json = (
		"config" => \%config,
		"audio"  => \%audio,
	);

	my %headers =('Content-Type' => "application/json");
	my %options = ('headers' => \%headers, 'content' => encode_json(\%json));
	my $response = $http->request('POST', "$url:recognize?key=$key", \%options);
	if (!$response->{'success'}) {
		say_msg("Failed to get data for file: $file");
		++$error;
		next;
	}
	if ($output eq "raw") {
		print $response->{'content'};
		next;
	}
	my $jdata = decode_json($response->{'content'});
	if ($output eq "detailed") {
		foreach (@{$jdata->{"results"}[0]->{"alternatives"}}) {
			printf "%-10s : %s\n", "transcript", $_->{"transcript"};
			printf "%-10s : %s\n", "confidence", $_->{"confidence"} if $_->{"confidence"};
		}
	} elsif ($output eq "compact") {
		print $_->{"transcript"}."\n" foreach (@{$jdata->{"results"}[0]->{"alternatives"}});
	}
}

exit(($error) ? 1 : 0);

sub parse_options {
# Command line options parsing #
	if (defined $options{k}) {
	# check API key #
		$key = $options{k};
	} else {
		say_msg("Invalid or missing API key.\n");
		exit 1;
	}
	if (defined $options{l}) {
	# check if language setting is valid #
		if ($options{l} =~ /^[a-z]{2}(-[a-zA-Z]{2,6})?$/) {
			$language = $options{l};
		} else {
			say_msg("Invalid language setting. Using default.\n");
		}
	}
	if (defined $options{o}) {
	# check if output setting is valid #
		if ($options{o} =~ /^(detailed|compact|raw)$/) {
			$output = $options{o};
		} else {
			say_msg("Invalid output formatting setting. Using default.\n");
		}
	}
	if (defined $options{n}) {
	# set number or results #
		$results = $options{n} if ($options{n} =~ /\d+/);
	}
	if (defined $options{r}) {
	# set audio sampling rate #
		$samplerate = $options{r} if ($options{r} =~ /\d+/);
	}
	# set profanity filter #
	$pro_filter = "true" if (defined $options{f});

	return;
}

sub encode_flac {
# Encode file to flac and return the filename #
	my $file   = shift;
	my $tmpdir = "/tmp";
	if (!$flac) {
		$flac   = `/usr/bin/which flac`;
		if (!$flac) {
			say_msg("flac encoder is missing. Aborting.");
			return -1;
		}
		chomp($flac);
	}
	my ($fh, $tmpname) = tempfile(
		"recg_XXXXXX",
		DIR    => $tmpdir,
		SUFFIX => '.flac',
		UNLINK => 1,
	);
	if (system($flac, "-1", "-f", "--totally-silent", "-o", "$tmpname", "$file")) {
		say_msg("$flac failed to encode file");
		return -1;
	}
	return $tmpname;
}

sub say_msg {
# Print messages to user if 'quiet' flag is not set #
	my @message = @_;
	warn @message if (!defined $options{q});
	return;
}

sub VERSION_MESSAGE {
# Help message #
	print "Speech recognition using Google Cloud Speech API.\n\n",
		"Usage: $0 [options] [file(s)]\n\n",
		"Supported options:\n",
		" -k <key>       specify the Speech API key\n",
		" -l <lang>      specify the language to use (default 'en-US')\n",
		" -o <type>      specify the type of output formatting\n",
		"    detailed    print detailed output with info like confidence (default)\n",
		"    compact     print only the transcripted string\n",
		"    raw         raw JSON output\n",
		" -r <rate>      specify the audio sample rate in Hz (default 16000)\n",
		" -n <number>    specify the maximum number of results (default 1)\n",
		" -f             filter out profanities\n",
		" -q             don't print any error messages or warnings\n",
		" -h             this help message\n\n";
	exit(1);
}
